src/porepy/numerics/ad/secondary_operator.py

```diff
@@ -0,0 +1,897 @@
+
+    @boundary_values.setter
+    def boundary_values(self, val: np.ndarray) -> None:
+
+        i = 0
+        for grid in self.mdg.boundaries():
+            n = self.num_dofs_on_grid(grid)
+            self.update_boundary_values(val[i : i + n], grid)
+            i += n
+
+    @property
+    def subdomain_values(self) -> np.ndarray:
+        """Property to access and store the value at the current time step, current
+        iterate on **all** subdomains in the md-grid for convenience.
+
+        The getter fetches the values in the order imposed by the md-grid.
+        The setter shifts iterate values backwards and sets the given value as the most
+        recent iterate.
+
+        Parameters:
+            val: A new value with shape ``(N,)`` to be set. Note that ``N`` is
+                calculated from the information passed in ``dof_info`` during
+                instantiation, and the number of cells, faces and nodes in all
+                subdomains.
+
+        Returns:
+            The current iterate value of same shape as ``val``.
+
+        """
+        vals = []
+        for _, data in self.mdg.subdomains(return_data=True):
+            vals.append(pp.get_solution_values(self.name, data, iterate_index=0))
+        if len(vals) > 0:
+            return np.hstack(vals)
+        else:
+            return np.zeros(0, dtype=float)
+
+    @subdomain_values.setter
+    def subdomain_values(self, val: np.ndarray) -> None:
+
+        i = 0
+        for grid in self.mdg.subdomains():
+            n = self.num_dofs_on_grid(grid)
+            self.progress_iterate_values_on_grid(val[i : i + n], grid)
+            i += n
+
+    @property
+    def subdomain_derivatives(self) -> np.ndarray:
+        """Property to access and store the derivatives at the current iteration of the
+        current time step, on **all** subdomains in the md grid for convenience.
+
+        The derivatives values are stored row-wise per dependency, column wise per
+        subdomain.
+
+        The getter fetches the values in the order imposed by the md-grid.
+        The setter shifts iterate values backwards and sets the given value as the
+        current iterate.
+
+        Important:
+            The order of derivatives should reflect the order of ``dependencies``
+            passed at instantiation.
+
+        Parameters:
+            val: ``shape=(num_dependencies, N)``
+
+                A new value to be set. Note that ``N`` must be consistent with the shape
+                of values for :meth:`subdomain_values`.
+
+        Returns:
+            The current iterate derivative values of same shape as ``val``.
+
+        """
+        vals = []
+        for _, data in self.mdg.subdomains(return_data=True):
+            vals.append(pp.get_solution_values(self._name_diffs, data, iterate_index=0))
+        if len(vals) > 0:
+            return np.hstack(vals)
+        else:
+            return np.zeros(self.num_dependencies, dtype=float)
```

Comment:
What happens here? What would the else correspond to?

Reply:
Yes, explicit raising of errors are usually preferable.

Reply:
A generic check for type and shape of values to be set is implemented.
Note: currently we allow arrays which are "too large" in the global setters. The grid-wise slicing allows. Should we add errors requiring the new values to be of exact shape, and not "at least of shape" for the global setters?

Reply:
>  Should we add errors requiring the new values to be of exact shape, and not "at least of shape" for the global setters?

I think we can do without this. Feel free to resolve the comment if you consider yourself done here.